blocksuite/affine/widget-toolbar/src/toolbar.ts

```diff
@@ -0,0 +1,424 @@
+          const viewSelection = result?.data;
+          activated = Boolean(viewSelection && viewSelection.type === 'area');
+        }
+      }
+
+      batch(() => {
+        flags.toggle(Flag.Native, activated);
+
+        if (!activated) return;
+
+        range$.value = activated ? range : null;
+        flavour$.value = 'affine:note';
+
+        flags.refresh(Flag.Native);
+      });
+    });
+
+    // Selects blocks in note.
+    disposables.add(
+      std.selection.filter$(BlockSelection).subscribe(result => {
+        const count = result.length;
+        let flavour = 'affine:note';
+        let activated = context.activated && Boolean(count);
+
+        if (activated) {
+          // Handles a signal block.
+          const block = count === 1 && std.store.getBlock(result[0].blockId);
+
+          // Chencks if block's config exists.
+          if (block) {
+            const modelFlavour = block.model.flavour;
+            const existed =
+              toolbarRegistry.modules.has(modelFlavour) ||
+              toolbarRegistry.modules.has(`custom:${modelFlavour}`);
+            if (existed) {
+              flavour = modelFlavour;
+            } else {
+              activated = matchModels(block.model, [
+                ParagraphBlockModel,
+                ListBlockModel,
+                CodeBlockModel,
+                ImageBlockModel,
+              ]);
+            }
+          }
+        }
+
+        batch(() => {
+          flavour$.value = flavour;
+
+          flags.toggle(Flag.Block, activated);
+
+          if (!activated) return;
+
+          flags.refresh(Flag.Block);
+        });
+      })
+    );
+
+    // Selects elements in edgeless.
+    // Triggered only when not in editing state.
+    disposables.add(
+      std.selection.filter$(SurfaceSelection).subscribe(result => {
+        const activated =
+          context.activated &&
+          Boolean(result.length) &&
+          !result.some(e => e.editing);
+        flags.toggle(Flag.Surface, activated);
+      })
+    );
+
+    disposables.add(
+      std.selection.slots.changed.on(results => {
+        if (!context.activated) return;
+
+        const value = flags.value$.peek();
+        if (flags.contains(Flag.Hovering | Flag.Hiding, value)) return;
+        if (!flags.check(Flag.Text, value)) return;
+
+        const hasTextSelection =
+          results.filter(s => s.is(TextSelection)).length > 0;
+        if (!hasTextSelection) return;
+
+        const range = std.range.value ?? null;
+        range$.value = range && !range.collapsed ? range : null;
+
+        // TODO(@fundon): maybe here can be further optimized
+        // 1. Prevents flickering effects.
+        // 2. We cannot use `host.getUpdateComplete()` here
+        // because it would cause excessive DOM queries, leading to UI jamming.
+        nextTick()
+          .then(() => flags.refresh(Flag.Text))
+          .catch(console.error);
+      })
```

Comment:
How is this different from the first subscription of `TextSelection`?

Reply:
https://github.com/toeverything/AFFiNE/blob/59de4558bdda4fc52974f610049ee3999663a9a8/blocksuite/framework/store/src/extension/selection/selection-extension.ts#L163-L169

https://github.com/toeverything/AFFiNE/blob/59de4558bdda4fc52974f610049ee3999663a9a8/blocksuite/framework/store/src/extension/selection/selection-extension.ts#L66-L73

`slots.changed` will notify the update directly.
The subscription of `TextSelection` needs to wait for awareness notification.

Reply:
I think some optimization is needed here, maybe `slots.changed` is a better option.